### ISSUE TITLE
pmdaprometheus issues

### ISSUE THREAD

**Issue by fche** [performancecopilot/pcp#343]
Found a couple of problems while trying to put the code to use; will add more as encountered.

- there appears to be no auto-configuration (issue #308) capability where a list of URLs (only) is supplied, and from which the pmda can over time efficiently collect/generate metadata and data

- the parsing code is fragile.  It can't consume #PCP comments of the form that modern pmwebd adds; it can't tolerate blank lines that the prometheus exposition-formats web page sample itself has; it does not appear to have been tested against a diverse set of prometheus exporters

**Comment by goodwinos** [comment]
- Another issue: qa/1192 requires the prometheus_client module but there is no RPM package available on RHEL or Fedora (it would be called python-prometheus_client or python3-prometheus_client). So that's a broken dep.
- Workaround (on F26 at least) is to: pip install prometheus-client; pip3 install prometheus-client

- The PMDA also uses OrderedDict, which isn't supported on older RHEL6/Fedora, but that's probably Ok since we'd only really use this on RHEL7 or later.

**Comment by goodwinos** [comment]
I might add: it's not really a broken dep. Only the qa tests in the pmda.prometheus group need the prometheus_client module (for the prometheus_client_server.python program, as used by the qa tests). The PMDA itself doesn't import that module.

**Comment by jronak** [comment]
@fche You are right, the parsing code does not handle # and blank lines. Thanks for pointing out.

@goodwinos I can get rid of the prometheus_client, would that help ?. 
Also Ordered Dict is already used by pmdaJSON and pmrep. What's the py version supported on RHEL6 ?

**Comment by myllynen** [comment]
OrderedDict is available with Python 2.7+ so all ok for RHEL 7+:

https://docs.python.org/2/library/collections.html
https://docs.python.org/3/library/collections.html

The first Fedora release with Python 2.7 was released way back in 2010 so we can ignore Fedora releases without OrderedDict support:

https://fedoraproject.org/wiki/Features/Python_2.7

RHEL 6 is a special case, it comes with Python 2.6 (which in upstream does not have OrderedDict support) but for RHEL 6 it was added in RHEL 6.5 so all supported RHEL 6 minor releases now have OrderedDict:

https://access.redhat.com/support/policy/updates/errata/

RHEL 5 is unsupported so can be ignored.

Python 2.7 is available at least since Ubuntu 14.04 and openSUSE 11.4 which both are already unsupported.

In summary, OrderedDict is supported on all relevant distributions / versions.

**Comment by fche** [comment]
current pmdaprometheus solves the above issues
